## Taller N°1 : Aplicacion de Algoritmos de B´usqueda

* Profesor: Jorge Ivan Padilla Buritica
* Estudiante: Juan Felipe Cardona Arango 
* Maestria en ciencia de datos y analitica
* Curso Fundamentos de Inteligencia Artificial
* 2025-02


## Preparacion del entorno de trabajo

In [21]:
import math
import heapq
import time
from time import time

# Punto 1

## Pseudocodigo

In [24]:
\begin{algorithm}[H]
\caption{A\* (Búsqueda con Heurística)}
\begin{algorithmic}[1]
\STATE \textbf{Entrada:} Grafo $G$, nodo inicio, nodo meta, heurística $h$
\STATE Inicializar:
\STATE\quad $\mathit{frontera} \leftarrow$ \texttt{ColaDePrioridad()} \COMMENT{almacena (nodo, $f=g+h$)}
\STATE\quad $g[\text{inicio}]\leftarrow 0$
\STATE\quad $\mathit{frontera}.\text{insertar}(\text{inicio},h(\text{inicio}))$
\STATE\quad $\mathit{padre}\leftarrow \{\}$

\WHILE{$\mathit{frontera}$ no esté vacía}
  \STATE $n \leftarrow \mathit{frontera}.\text{extraer\_min}()$
  \IF{$n = \text{meta}$}
    \RETURN \texttt{reconstruir\_camino}($\mathit{padre},n$)
  \ENDIF
  \FORALL{vecino en $G.\text{adyacentes}(n)$}
    \STATE $g_{\text{tentativo}}\leftarrow g[n]+\text{costo}(n,vecino)$
    \IF{vecino no en $g$ \OR $g_{\text{tentativo}}<g[\text{vecino}]$}
      \STATE $g[\text{vecino}]\leftarrow g_{\text{tentativo}}$
      \STATE $f\leftarrow g_{\text{tentativo}}+h(\text{vecino})$
      \STATE $\mathit{frontera}.\text{insertar\_o\_actualizar}(\text{vecino},f)$
      \STATE $\mathit{padre}[\text{vecino}]\leftarrow n$
    \ENDIF
  \ENDFOR
\ENDWHILE
\RETURN \texttt{fallo}
\end{algorithmic}
\end{algorithm}


SyntaxError: unexpected character after line continuation character (2552479889.py, line 1)

## Codigo

In [11]:
class Node:
    def __init__(self, id, coord):
        self.id = id          # Identificador único
        self.coord = coord    # (x, y) en coordenadas rurales
        self.edges = []       # Lista de tuplas (vecino, costo)

    def add_edge(self, neighbor, cost):
        self.edges.append((neighbor, cost))

# Heurística: distancia euclidiana
def heuristic(a: Node, b: Node) -> float:
    (x1, y1), (x2, y2) = a.coord, b.coord
    return math.hypot(x2 - x1, y2 - y1)

# A* Search
def astar(graph, start: Node, goal: Node):
    """A* Search: graph no dirigido implícito en conexiones de nodos."""
    t0 = time.time()
    # Aunque recibe graph por compatibilidad con compare_algorithms, no se utiliza explícitamente.
    frontier = [(0 + heuristic(start, goal), 0, start, None)]
    # (f = g + h, g = coste acumulado, nodo actual, padre)
    visited = {}
    expanded = 0

    while frontier:
        f, g, current, parent = heapq.heappop(frontier)
        expanded += 1
        if current.id in visited:
            continue
        visited[current.id] = (parent, g)
        if current.id == goal.id:
            break
        for neighbor, cost in current.edges:
            if neighbor.id in visited:
                continue
            g2 = g + cost
            f2 = g2 + heuristic(neighbor, goal)
            heapq.heappush(frontier, (f2, g2, neighbor, current.id))

    # Reconstrucción de la ruta óptima
    path, node = [], goal.id
    while node is not None:
        path.append(node)
        node, _ = visited[node]
    path = list(reversed(path))
    t1 = time.time()

    return {
        'path': path,
        'cost': visited[goal.id][1],
        'time': t1 - t0,
        'expanded': expanded,
        'depth': len(path) - 1
    }

## Implementacion

In [12]:
def bfs(graph, start, goal):
    """Breadth-First Search stub: implementar búsqueda en anchura."""
    # TODO: implementar BFS que retorne:
    # {'path': [...], 'cost': ..., 'expanded': ..., 'depth': ...}
    return {'path': [], 'cost': None, 'expanded': 0, 'depth': 0}

def dfs(graph, start, goal):
    """Depth-First Search stub: implementar búsqueda en profundidad."""
    # TODO: implementar DFS con la misma estructura de retorno
    return {'path': [], 'cost': None, 'expanded': 0, 'depth': 0}

def ucs(graph, start, goal):
    """Uniform Cost Search stub: implementar búsqueda de costo uniforme."""
    # TODO: implementar UCS similar a A* sin heurística
    return {'path': [], 'cost': None, 'expanded': 0, 'depth': 0}

def compare_algorithms(graph, start, goal, algorithms):
    results = {}
    for name, func in algorithms.items():
        t0 = time()
        res = func(graph, start, goal)
        res['time'] = time() - t0
        results[name] = res
    return results


In [ ]:
import time
import heapq
from collections import deque

# Breadth-First Search
def bfs(graph, start, goal):
    """Breadth-First Search: camino mínimo en número de aristas."""
    t0 = time.time()
    queue = deque([start])
    visited = {start.id: (None, 0)}  # nodo_id: (padre_id, depth)
    expanded = 0

    while queue:
        current = queue.popleft()
        expanded += 1
        if current.id == goal.id:
            break
        for neighbor, cost in current.edges:
            if neighbor.id not in visited:
                visited[neighbor.id] = (current.id, visited[current.id][1] + 1)
                queue.append(neighbor)

    # Reconstrucción de la ruta
    path, node_id = [], goal.id
    node_map = {n.id: n for n in graph}
    while node_id is not None:
        path.append(node_id)
        node_id, _ = visited[node_id]
    path = list(reversed(path))

    # Cálculo del costo real según pesos
    cost = 0
    for i in range(len(path) - 1):
        node = node_map[path[i]]
        for nb, c in node.edges:
            if nb.id == path[i+1]:
                cost += c
                break

    t1 = time.time()
    return {'path': path, 'cost': cost, 'expanded': expanded, 'depth': len(path) - 1, 'time': t1 - t0}

# Depth-First Search
def dfs(graph, start, goal):
    """Depth-First Search: explora en profundidad, devuelve primer camino encontrado."""
    t0 = time.time()
    visited = {}
    expanded = 0
    result = None

    def dfs_rec(node, parent_id, depth):
        nonlocal expanded, result
        if node.id in visited or result:
            return
        visited[node.id] = (parent_id, depth)
        expanded += 1
        if node.id == goal.id:
            # Reconstrucción de la ruta
            path, nid = [], node.id
            while nid is not None:
                path.append(nid)
                nid, _ = visited[nid]
            result = path[::-1]
            return
        for neighbor, cost in node.edges:
            dfs_rec(neighbor, node.id, depth + 1)

    dfs_rec(start, None, 0)
    path = result or []
    t1 = time.time()
    return {'path': path, 'cost': None, 'expanded': expanded, 'depth': len(path) - 1 if path else None, 'time': t1 - t0}

# Uniform Cost Search
def ucs(graph, start, goal):
    """Uniform Cost Search: encuentra camino óptimo según pesos."""
    t0 = time.time()
    frontier = [(0, start, None)]  # (g, nodo, padre)
    visited = {}
    expanded = 0

    while frontier:
        g, current, parent = heapq.heappop(frontier)
        if current.id in visited:
            continue
        visited[current.id] = (parent.id if parent else None, g)
        expanded += 1
        if current.id == goal.id:
            break
        for neighbor, cost in current.edges:
            if neighbor.id not in visited:
                heapq.heappush(frontier, (g + cost, neighbor, current))

    # Reconstrucción de la ruta
    path, nid = [], goal.id
    while nid is not None:
        path.append(nid)
        nid, _ = visited[nid]
    path = path[::-1]
    t1 = time.time()
    return {'path': path, 'cost': visited[goal.id][1], 'expanded': expanded, 'depth': len(path) - 1, 'time': t1 - t0}

# Comparador de algoritmos
def compare_algorithms(graph, start, goal, algorithms):
    """Ejecuta y compara múltiples algoritmos, retornando métricas por algoritmo."""
    results = {}
    for name, func in algorithms.items():
        results[name] = func(graph, start, goal)
    return results



{'BFS': {'path': ['A', 'B', 'C'], 'cost': 3.6500000000000004, 'expanded': 3, 'depth': 2, 'time': 0.0}, 'DFS': {'path': ['A', 'B', 'C'], 'cost': None, 'expanded': 3, 'depth': 2, 'time': 0.0}, 'UCS': {'path': ['A', 'B', 'C'], 'cost': 3.6500000000000004, 'expanded': 3, 'depth': 2, 'time': 0.0014858245849609375}, 'A*': {'path': ['A', 'B', 'C'], 'cost': 3.6500000000000004, 'time': 0.0, 'expanded': 3, 'depth': 2}}


In [20]:
# Ejemplo de uso:
node_A = Node('A', (0, 0))
node_B = Node('B', (1, 2))
node_C = Node('C', (2, 1))
node_A.add_edge(node_B, 2.24)
node_B.add_edge(node_A, 2.24)
node_B.add_edge(node_C, 1.41)
node_C.add_edge(node_B, 1.41)
grafo = [node_A, node_B, node_C]
nodo_inicio = node_A
nodo_destino = node_C
algorithms = {'BFS': bfs, 'DFS': dfs, 'UCS': ucs, 'A*': astar}
metrics = compare_algorithms(grafo, nodo_inicio, nodo_destino, algorithms)
print(metrics)

{'BFS': {'path': ['A', 'B', 'C'], 'cost': 3.6500000000000004, 'expanded': 3, 'depth': 2, 'time': 0.0}, 'DFS': {'path': ['A', 'B', 'C'], 'cost': None, 'expanded': 3, 'depth': 2, 'time': 0.0}, 'UCS': {'path': ['A', 'B', 'C'], 'cost': 3.6500000000000004, 'expanded': 3, 'depth': 2, 'time': 0.0}, 'A*': {'path': ['A', 'B', 'C'], 'cost': 3.6500000000000004, 'time': 0.0, 'expanded': 3, 'depth': 2}}


## Interpretación de resultados y comparación de algoritmos
A continuación se comentan las métricas obtenidas en el ejemplo para cada algoritmo:

- **BFS**:
  - *Path*: `['A', 'B', 'C']` coincide con el camino de menor número de saltos.
  - *Cost*: 3.65 (suma de pesos de aristas). Aunque BFS no utiliza pesos para decidir la ruta, aquí se calcula el coste real tras la búsqueda.
  - *Expanded*: 3 nodos (todos los nodos fueron visitados hasta alcanzar el destino).
  - *Depth*: 2 aristas.
  - *Time*: ~0s (grafos pequeños).

- **DFS**:
  - *Path*: `['A', 'B', 'C']` primer camino encontrado en profundidad.
  - *Cost*: no aplicable (se omitió cálculo de coste en DFS, pero podría añadirse como en BFS).
  - *Expanded*: 3 nodos.
  - *Depth*: 2.
  - *Time*: ~0s.

- **UCS**:
  - *Path*: `['A', 'B', 'C']` ruta óptima ponderada.
  - *Cost*: 3.65 (optimo según pesos).
  - *Expanded*: 3 nodos.
  - *Depth*: 2.
  - *Time*: ~0s.

- **A***:
  - *Path*: `['A', 'B', 'C']` misma ruta óptima.
  - *Cost*: 3.65.
  - *Expanded*: 3 nodos (igual que UCS en este grafo pequeño y heurística perfecta).
  - *Depth*: 2.
  - *Time*: ~0s.

**Reflexiones**:
1. Los cuatro algoritmos hallan la misma ruta en este escenario trivial.
2. BFS y DFS no optimizan en coste, pero BFS encuentra el camino de menor longitud en aristas.
3. UCS y A* garantizan optimalidad en coste; A* expande igual que UCS aquí porque la heurística euclidiana coincide con costes reales.
4. En grafos mayores con variantes de pesos y geometría más compleja, se observará diferencia en número de nodos expandidos y tiempo de cómputo.
5. Se recomienda realizar comparaciones en escenarios más grandes y heterogéneos para evidenciar ventajas de la heurística en A* frente a UCS.



## Respuestas a Preguntas Orientadoras

# Punto 2

## Pseudocodigo 

ALGORITMO Anchura(G, inicio, meta)
    frontera ← Cola()                    // FIFO
    frontera.encolar(inicio)
    explorados ← {inicio}
    padre ← diccionario vacío

    MIENTRAS frontera NO esté vacía HACER
        n ← frontera.desencolar()
        SI n = meta ENTONCES
            DEVOLVER reconstruir_camino(padre, n)
        FIN SI

        PARA CADA vecino EN G.adyacentes(n) HACER
            SI vecino NO está EN explorados ENTONCES
                explorados.agregar(vecino)
                frontera.encolar(vecino)
                padre[vecino] ← n
            FIN SI
        FIN PARA
    FIN MIENTRAS

    DEVOLVER fallo
FIN ALGORITMO

In [ ]:
## Codigo 

In [ ]:
# Punto 3

## Pseudocodigo

ALGORITMO Anchura(G, inicio, meta)
    // Inicialización
    frontera ← Cola()                    // FIFO
    frontera.encolar(inicio)
    explorados ← {inicio}
    padre ← diccionario vacío

    // Bucle principal
    MIENTRAS frontera NO esté vacía HACER
        n ← frontera.desencolar()
        SI n = meta ENTONCES
            DEVOLVER reconstruir_camino(padre, n)
        FIN SI

        // Expandir vecinos no explorados
        PARA CADA vecino EN G.adyacentes(n) HACER
            SI vecino NO está EN explorados ENTONCES
                explorados.agregar(vecino)
                frontera.encolar(vecino)
                padre[vecino] ← n
            FIN SI
        FIN PARA
    FIN MIENTRAS

    // Si no se encuentra la meta
    DEVOLVER fallo
FIN ALGORITMO




In [ ]:
## Codigo

## Respuestas a Preguntas Orientadoras

# Punto 3

## Pseudocodigo

SUBALGORITMO Profundidad_Limitada(G, n, meta, límite)
    SI n = meta ENTONCES
        DEVOLVER reconstruir_camino(global_padre, n)
    FIN SI
    SI límite = 0 ENTONCES
        DEVOLVER "corte"                           // profundidad agotada
    FIN SI
    corte_ocurrió ← FALSO
    PARA CADA vecino EN G.adyacentes(n) HACER
        SI vecino NO visitado ENTONCES
            global_padre[vecino] ← n
            resultado ← Profundidad_Limitada(G, vecino, meta, límite-1)
            SI resultado = "corte" ENTONCES
                corte_ocurrió ← VERDADERO
            SINO SI resultado ≠ fallo ENTONCES
                DEVOLVER resultado                // se halló solución
            FIN SI
        FIN SI
    FIN PARA
    SI corte_ocurrió ENTONCES DEVOLVER "corte" SINO DEVOLVER fallo
FIN SUBALGORITMO


ALGORITMO ACMR_Profundidad_Limitada(Especie a, Especie b, límite L, DAG filogenia)
    ancestros_a ← Ancestros_Limitados(a, L, filogenia)
    ancestros_b ← Ancestros_Limitados(b, L, filogenia)
    comunes ← intersección(ancestros_a, ancestros_b)
    SI comunes vacía ENTONCES DEVOLVER "sin ancestro común dentro de L"
    DEVOLVER más_profundo(comunes)                 // profundidad mínima desde a y b
FIN ALGORITMO


## Codigo

## Respuestas a Preguntas Orientadoras

# Punto 4

## Pseudocodigo

ALGORITMO ACMR_Prof_Lim(a, b, L, grafo)
    // Obtener ancestros hasta L niveles
    A ← Ancestros_Limitados(a, L, grafo)
    B ← Ancestros_Limitados(b, L, grafo)

    comunes ← A ∩ B
    SI comunes vacío ENTONCES
        DEVOLVER "sin ancestro común hasta L"
    FIN SI

    DEVOLVER ancestro_más_profundo(comunes)
FIN ALGORITMO

SUBALGORITMO Ancestros_Limitados(nodo, L, grafo)
    conjunto ← {nodo}
    SI L = 0 ENTONCES
        DEVOLVER conjunto
    FIN SI

    PARA CADA padre EN grafo.predecesores(nodo) HACER
        conjunto ← conjunto ∪ Ancestros_Limitados(padre, L-1, grafo)
    FIN PARA

    DEVOLVER conjunto
FIN SUBALGORITMO


## Codigo

## Respuestas a Preguntas Orientadoras